src/extent.c

```diff
@@ -1150,45 +1170,37 @@ extent_grow_retained(tsdn_t *tsdn, arena_t *arena,
 			memset(addr, 0, size);
 		}
 	}
-	/*
-	 * Increment extent_grow_next, but take care to do so atomically and
-	 * bail out if the increment would exceed the legal range.
-	 */
-	pszind_t egn = atomic_load_u(&arena->extent_grow_next, ATOMIC_RELAXED);
-	while (true) {
-		if (egn + 1 == NPSIZES) {
-			break;
-		}
-		assert(egn + 1 < NPSIZES);
-		if (atomic_compare_exchange_weak_u(&arena->extent_grow_next,
-		    &egn, egn + 1, ATOMIC_RELAXED, ATOMIC_RELAXED)) {
-			break;
-		}
-	}
+
 	return extent;
+label_err:
+	malloc_mutex_unlock(tsdn, &arena->extent_grow_next_mtx);
+	return NULL;
 }
 
 static extent_t *
 extent_alloc_retained(tsdn_t *tsdn, arena_t *arena,
     extent_hooks_t **r_extent_hooks, void *new_addr, size_t size, size_t pad,
     size_t alignment, bool slab, szind_t szind, bool *zero, bool *commit) {
-	extent_t *extent;
-
 	assert(size != 0);
 	assert(alignment != 0);
 
-	extent = extent_recycle(tsdn, arena, r_extent_hooks,
+	malloc_mutex_lock(tsdn, &arena->extent_grow_next_mtx);
```

Comment:
Is it true that this doesn't cause much additional contention because we need the extents lock in extent_recycle anyway (and grow rarely happen)?